clippy_lints/src/expr_metavars_in_unsafe.rs

```diff
@@ -0,0 +1,241 @@
+use std::collections::btree_map::Entry;
+use std::collections::BTreeMap;
+
+use clippy_utils::diagnostics::span_lint_hir_and_then;
+use clippy_utils::is_lint_allowed;
+use itertools::Itertools;
+use rustc_hir::intravisit::{walk_block, walk_expr, walk_stmt, Visitor};
+use rustc_hir::{BlockCheckMode, Expr, ExprKind, HirId, Stmt, UnsafeSource};
+use rustc_lint::{LateContext, LateLintPass};
+use rustc_session::impl_lint_pass;
+use rustc_span::def_id::DefId;
+use rustc_span::{sym, Span, SyntaxContext};
+
+declare_clippy_lint! {
+    /// ### What it does
+    /// Looks for macros that expand metavariables in an unsafe block.
+    ///
+    /// ### Why is this bad?
+    /// This is unsound: it allows the user of the macro to write unsafe code outside of an
+    /// unsafe block at callsite, potentially invoking undefined behavior in safe code.
+    ///
+    /// ### Known limitations
+    /// Due to how macros are represented in the compiler at the time Clippy runs its lints,
+    /// it's not possible to look for metavariables in macro definitions directly.
+    ///
+    /// Instead, this lint looks at expansions of macros defined in the same crate.
+    /// This leads to false negatives when a macro is never actually invoked.
+    ///
+    /// ### Example
```

Comment:
Since the lint explicitly checks for cases, where the metavariable is also used in safe and unsafe blocks, I wonder if that could be a valid suggestion. It shouldn't be the default, but if someone really needs the unsafe block at this location for some reason, they could just add a `if false {$expr}` somewhere.

I'm considering if we might want to mention this in the lint description, but it feels like that's just a hack and not a nice solution. If you agree, you can just resolve this :D

Reply:
I'm not a big fan of that hack but yeah I suppose it is worth mentioning in the description anyway as a last resort if there is no other way for people to restructure their macros. Added to the documentation.